In [1]:
import pandas as pd
from sqlalchemy import create_engine, text
import urllib.parse

## Database Connectivity

In [2]:
username = 'root'
password = 'khan@240405'  
database_name = 'zomato'  
encoded_pwd = urllib.parse.quote_plus(password)

In [3]:
connection_string = f'mysql+pymysql://{username}:{encoded_pwd}@localhost:3306/{database_name}'

In [4]:
engine = create_engine(connection_string)

In [5]:
pd.read_sql('show tables',con=engine)

Tables_in_zomato
0             food
1   loyalcustomers
2             menu
3     orderdetails
4           orders
5          partner
6            resto
7            users

In [6]:
table_name = ['food', 'menu', 'orderdetails', 'orders', 'partner', 'resto', 'users']

In [7]:
df = {}

In [8]:
for table in table_name:
    df[table] = pd.read_sql(f'select * from {table}',con=engine)
     

In [9]:
df['orders']

orderid  userid  rid  amount  partnerid  deliverytime  deliveryrating  \
0      1001       1    1     550          1            25               5   
1      1002       1    2     415          1            19               5   
2      1003       1    3     240          5            29               4   
3      1004       1    3     240          4            42               3   
4      1005       1    3     220          1            58               1   
5      1006       2    1     950          2            16               5   
6      1007       2    2     530          3            60               1   
7      1008       2    3     240          5            33               4   
8      1009       2    4     300          4            41               1   
9      1010       2    5     650          1            67               1   
10     1011       3    1     450          2            25               3   
11     1012       3    4     180          5            33               4   
12     1013       3    2     230          4            45               3   
13     1014       3    2     230          2            55               1   
14     1015       3    2     230          3            21               5   
15     1016       4    4     300          3            31               5   
16     1017       4    4     300          1            50               1   
17     1018       4    4     400          2            40               3   
18     1019       4    5     400          1            70               2   
19     1020       4    5     400          3            26               5   
20     1021       5    1     550          5            22               2   
21     1022       5    1     550          1            34               5   
22     1023       5    2     645          4            38               5   
23     1024       5    2     645          2            58               2   
24     1025       5    2     645          2            44               4   

    restorating        date  
0             3  2022-05-10  
1             2  2022-05-26  
2             0  2022-06-15  
3             5  2022-06-29  
4             4  2022-07-10  
5             0  2022-06-10  
6             5  2022-06-23  
7             5  2022-07-07  
8             0  2022-07-17  
9             4  2022-07-31  
10            1  2022-05-10  
11            1  2022-05-20  
12            0  2022-05-30  
13            2  2022-06-11  
14            0  2022-06-22  
15            5  2022-05-15  
16            0  2022-05-30  
17            5  2022-06-15  
18            4  2022-06-30  
19            3  2022-07-15  
20            0  2022-07-01  
21            1  2022-07-08  
22            1  2022-07-15  
23            1  2022-07-21  
24            0  2022-07-28

In [10]:
for table, val in df.items():
    print(f"Data from table '{table}':")
    print(val.head()) 
    print('-'*100)

Data from table 'food':
   fid            fname     type
0    1    Non-veg Pizza  Non-veg
1    2        Veg Pizza      Veg
2    3  Choco Lava cake      Veg
3    4    Chicken Wings  Non-veg
4    5  Chicken Popcorn  Non-veg
----------------------------------------------------------------------------------------------------
Data from table 'menu':
   menuid  rid  fid  price
0       1    1    1    450
1       2    1    2    400
2       3    1    3    100
3       4    2    3    115
4       5    2    4    230
----------------------------------------------------------------------------------------------------
Data from table 'orderdetails':
   id  orderid  fid
0   1     1001    1
1   2     1001    3
2   3     1002    4
3   4     1002    3
4   5     1003    6
----------------------------------------------------------------------------------------------------
Data from table 'orders':
   orderid  userid  rid  amount  partnerid  deliverytime  deliveryrating  \
0     1001       1    1     550    

# Data Exploration

#### -- 1. Find users who have never ordered

In [12]:
pd.read_sql("select u.name from users u left join orders o on u.userid = o.userid where o.userid is NULL", con=engine)

name
0  Anupama
1  Rishabh

In [13]:
pd.read_sql("select name from users where userid not in (select userid from orders)", con=engine)

name
0  Anupama
1  Rishabh

#### -- 2. Find average price per dish

In [16]:
pd.read_sql("Select f.fname, round(avg(m.price),2) as AvgPrice from food f join menu m on  m.fid = f.fid group by 1 order by 2 desc" , con=engine)

fname  AvgPrice
0      Non-veg Pizza    450.00
1          Veg Pizza    400.00
2    Chicken Popcorn    300.00
3      Chicken Wings    230.00
4   Schezwan Noodles    220.00
5          Rice Meal    213.33
6        Masala Dosa    180.00
7     Veg Manchurian    180.00
8          Roti meal    140.00
9          Rava Idli    120.00
10   Choco Lava cake     98.33

#### --3. Find top restaurant in terms of orders for a given month

In [52]:
pd.read_sql("describe orders",con=engine)

Field    Type Null Key Default Extra
0         orderid  bigint  YES        None      
1          userid  bigint  YES        None      
2             rid  bigint  YES        None      
3          amount  bigint  YES        None      
4       partnerid  bigint  YES        None      
5    deliverytime  bigint  YES        None      
6  deliveryrating  bigint  YES        None      
7     restorating  bigint  YES        None      
8            date    date  YES        None

In [19]:
pd.read_sql('''select r.name, count(*) as orderCounts from resto r join orders o on r.rid = o.rid 
where monthname(o.date) = 'June'
group by 1 order by 2 desc limit 1''',con=engine)

name  orderCounts
0  kfc            3

In [62]:
for table, val in df.items():
    print(f"Data from table '{table}':")
    print(val.head()) 
    print('-'*100)

Data from table 'food':
   fid            fname     type
0    1    Non-veg Pizza  Non-veg
1    2        Veg Pizza      Veg
2    3  Choco Lava cake      Veg
3    4    Chicken Wings  Non-veg
4    5  Chicken Popcorn  Non-veg
----------------------------------------------------------------------------------------------------
Data from table 'menu':
   menuid  rid  fid  price
0       1    1    1    450
1       2    1    2    400
2       3    1    3    100
3       4    2    3    115
4       5    2    4    230
----------------------------------------------------------------------------------------------------
Data from table 'orderdetails':
   id  orderid  fid
0   1     1001    1
1   2     1001    3
2   3     1002    4
3   4     1002    3
4   5     1003    6
----------------------------------------------------------------------------------------------------
Data from table 'orders':
   orderid  userid  rid  amount  partnerid  deliverytime  deliveryrating  \
0     1001       1    1     550    

#### -- 4. Restaurants with monthly sales > x in given month

In [20]:
pd.read_sql('''Select r.name, sum(o.amount) as revenue from resto r join orders o on r.rid = o.rid
            where monthname(o.date) = "June" 
             group by 1 having sum(o.amount) > 550''',con=engine)

name  revenue
0  dominos    950.0
1      kfc    990.0

#### -- 5. Show all orders with order details for a particular customer in a particular date.

In [21]:
pd.read_sql('''Select o.orderid, f.fid, f.fname, r.name from orders o join resto r on o.rid = r.rid 
join orderdetails od on od.orderid = o.orderid join food f on f.fid = od.fid 
where o.userid = (select userid from users where name = 'Ankit') and (date between '2022-06-10' and '2022-07-10')''',con=engine)

orderid  fid             fname        name
0     1019   10  Schezwan Noodles  China Town
1     1018   10  Schezwan Noodles  Dosa Plaza
2     1019   11    Veg Manchurian  China Town
3     1018   11    Veg Manchurian  Dosa Plaza

#### -- 6. Find restaurants with max repeated customers

In [26]:
pd.read_sql('''select o.name, count(*) as repeatedCustomers from (select rid, userid from orders group by 1,2 having count(*) > 1) t
            join resto o on o.rid = t.rid group by o.name order by 2 desc limit 1''', con=engine)

name  repeatedCustomers
0  kfc                  2

#### -- 7. Month over Month revenue growth of swiggy

In [11]:
pd.read_sql('''Select month, revenue, previousRevenue, round(((revenue-previousRevenue)/previousRevenue)*100,2) as MoM from (
with cte as (select monthname(date) as month, sum(amount) as revenue from orders group by 1) 
select month, revenue, lag(revenue,1) over (order by revenue) as previousRevenue from cte
) t ''',con=engine)

month  revenue  previousRevenue    MoM
0   May   2425.0              NaN    NaN
1  June   3220.0           2425.0  32.78
2  July   4845.0           3220.0  50.47

#### -- 8. Find Customer's fav food

In [13]:
pd.read_sql(''' with cte as (select userid, od.fid, count(*) as frequency from
orders o join orderdetails od on o.orderid = od.orderid group by 1,2)
select u.name, f.fname, c1.frequency from cte c1 join users u on u.userid = c1.userid join food f on f.fid = c1.fid
where frequency = (select max(frequency) from cte c2 where c1.userid = c2.userid) ''',con=engine)

name             fname  frequency
0      Neha   Choco Lava cake          5
1  Khushboo   Choco Lava cake          3
2    Nitish   Choco Lava cake          5
3   Vartika     Chicken Wings          3
4     Ankit  Schezwan Noodles          3
5     Ankit    Veg Manchurian          3

#### -- 9. Create a Loyal Customers Table with Users Who Have Placed More Than 3 Orders

In [17]:
with engine.connect() as conn:
    conn.execute(text('''Create table if not exists loyalcustomers(
    userid int primary key,
    name varchar(255),
    discount int(11)
    )
    '''))
    conn.commit()

In [18]:
pd.read_sql("show tables",con=engine)

Tables_in_zomato
0             food
1   loyalcustomers
2             menu
3     orderdetails
4           orders
5          partner
6            resto
7            users

In [19]:
with engine.connect() as conn:
    conn.execute(text(
        ''' insert into loyalcustomers (userid,name)
        select u.userid,u.name from users u join orders o on o.userid = u.userid group by 1,2 having count(*) > 3
        '''
    ))
    conn.commit()

In [20]:
df['loyalcustomers'] = pd.read_sql("select * from loyalcustomers",con=engine)

In [21]:
for table, val in df.items():
    print(f"Data from table {table} :")
    print(val.head())
    print('-'*100)

Data from table food :
   fid            fname     type
0    1    Non-veg Pizza  Non-veg
1    2        Veg Pizza      Veg
2    3  Choco Lava cake      Veg
3    4    Chicken Wings  Non-veg
4    5  Chicken Popcorn  Non-veg
----------------------------------------------------------------------------------------------------
Data from table menu :
   menuid  rid  fid  price
0       1    1    1    450
1       2    1    2    400
2       3    1    3    100
3       4    2    3    115
4       5    2    4    230
----------------------------------------------------------------------------------------------------
Data from table orderdetails :
   id  orderid  fid
0   1     1001    1
1   2     1001    3
2   3     1002    4
3   4     1002    3
4   5     1003    6
----------------------------------------------------------------------------------------------------
Data from table orders :
   orderid  userid  rid  amount  partnerid  deliverytime  deliveryrating  \
0     1001       1    1     550        

#### -- 10. Apply Discount for Loyal Customers Based on Their Order Value

In [27]:
with engine.connect() as conn:
    conn.execute(text(
        ''' update loyalcustomers
            set discount = (select sum(amount)*0.1 as discount from orders where loyalcustomers.userid = orders.userid)
            '''))
    conn.commit()

In [28]:
df['loyalcustomers'] = pd.read_sql("select * from loyalcustomers",con=engine)

In [29]:
df['loyalcustomers']

userid      name  discount
0       1    Nitish       167
1       2  Khushboo       267
2       3   Vartika       132
3       4     Ankit       180
4       5      Neha       304

#### -- 11. Most paired products 

In [50]:
pd.read_sql(""" select f1.fname as food1, f2.fname as food2, count(*) as frequency 
from orderdetails od1 join orderdetails od2 on od1.orderid = od2.orderid and od1.fid < od2.fid 
join food f1 on od1.fid = f1.fid join food f2 on f2.fid = od2.fid group by 1,2 having count(*) > 2 order by 3 desc""",con=engine)

food1            food2  frequency
0   Choco Lava cake    Chicken Wings          5
1     Non-veg Pizza  Choco Lava cake          4
2  Schezwan Noodles   Veg Manchurian          4
3   Choco Lava cake  Chicken Popcorn          3
4     Chicken Wings  Chicken Popcorn          3
5   Choco Lava cake        Rice Meal          3
6       Masala Dosa        Rava Idli          3

#### -- 12. List all food items and their corresponding restaurant names 

In [54]:
pd.read_sql(""" select distinct(f.fname), r.name from orders o join orderdetails od on od.orderid = o.orderid join food f on f.fid = od.fid 
join resto r on r.rid = o.rid""", con=engine)

fname        name
0      Non-veg Pizza     dominos
1          Veg Pizza     dominos
2    Choco Lava cake         kfc
3    Choco Lava cake     dominos
4    Choco Lava cake        box8
5      Chicken Wings         kfc
6    Chicken Popcorn         kfc
7          Rice Meal  China Town
8          Rice Meal        box8
9          Roti meal        box8
10       Masala Dosa  Dosa Plaza
11         Rava Idli  Dosa Plaza
12  Schezwan Noodles  China Town
13  Schezwan Noodles  Dosa Plaza
14    Veg Manchurian  China Town
15    Veg Manchurian  Dosa Plaza

#### -- 13. Find users who have placed orders at more than 3 different restaurants

In [66]:
pd.read_sql(""" select u.name from users u join orders o on o.userid = u.userid 
group by 1 having count(distinct(o.rid))>=3""",con=engine)

name
0  Khushboo
1    Nitish
2   Vartika

#### -- 14. Users Who Ordered at a Restaurant 3 or More Times

In [83]:
pd.read_sql(""" with cte as (select u.userid,u.name as userName, r.name as restoName, count(*) as frequency from users u 
join orders o on u.userid = o.userid join resto r on r.rid = o.rid group by 1,2,3)
select userName, restoName, frequency from cte c1 where frequency = (select max(frequency) as max from cte c2 where c1.userid = c2.userid)
and frequency >=3""",con=engine)

userName   restoName  frequency
0   Nitish        box8          3
1  Vartika         kfc          3
2    Ankit  Dosa Plaza          3
3     Neha         kfc          3

#### -- 15.  Identify the restaurant(s) with the highest average delivery time for orders in a specific month (e.g., June 2022).

In [91]:
pd.read_sql(""" select r.name as RestoName, round(avg(o.deliverytime),2) as AvgDeliveryTime from orders o 
join resto r on r.rid = o.rid where o.date >= '2022-06-01' AND o.date < '2022-07-01' group by 1  order by 2 desc limit 1 """,con=engine)

RestoName  AvgDeliveryTime
0  China Town             70.0

#### -- 16. Find the most popular cuisine among users based on the total number of orders placed.

In [93]:
pd.read_sql(""" select r.cuisine, count(o.orderid) as TotalOrdersPlaced from orders o 
join resto r on r.rid = o.rid group by 1 order by 2 desc limit 1 """,con=engine)

cuisine  TotalOrdersPlaced
0  American                  8

#### -- 17. Calculate the percentage contribution of each restaurant to the total revenue of Swiggy.

In [95]:
pd.read_sql(""" select r.name as RestoName, sum(o.amount) as RestoRevenue, round(sum(o.amount)/(select sum(amount) from orders) *100,2)
as PercentageContribution from orders o join resto r on r.rid = o.rid group by 1 order by 3 desc""",con=engine)

RestoName  RestoRevenue  PercentageContribution
0         kfc        3570.0                   34.03
1     dominos        3050.0                   29.08
2  Dosa Plaza        1480.0                   14.11
3  China Town        1450.0                   13.82
4        box8         940.0                    8.96

#### -- 18. Find users who have ordered both "Veg" and "Non-veg" items at least once.

In [107]:
pd.read_sql(""" select u.name as Name from users u join orders o on o.userid = u.userid join orderdetails od on od.orderid = o.orderid
join food f on f.fid = od.fid where f.type in ('Veg','Non-veg') group by 1 having count(distinct f.type) = 2 """,con=engine)

Name
0  Khushboo
1      Neha
2    Nitish
3   Vartika

#### -- 19. Identify the restaurant(s) with the lowest customer satisfaction (average restorating) but the highest sales.

In [110]:
pd.read_sql(""" with restoStats as (
select r.name as RestoName, avg(o.restorating) as AvgRating, sum(o.amount) as TotalSales 
from resto r join orders o on o.rid = r.rid group by 1
),
lowerCustomerSatisfaction as (
select RestoName, AvgRating, TotalSales from restoStats where AvgRating = (select min(AvgRating) from restoStats)
)
select RestoName, AvgRating, TotalSales from lowerCustomerSatisfaction order by TotalSales desc limit 1
""",con=engine)

RestoName  AvgRating  TotalSales
0   dominos        1.0      3050.0

#### -- 20. Find the restaurant(s) with the highest number of unique customers in a specific month.

In [125]:
pd.read_sql(""" select r.name, count(distinct o.userid) as UniqueCustomers from orders o join resto r on r.rid = o.rid 
where o.date between'2022-06-01' and '2022-07-01' group by 1 having count(distinct o.userid) = (
select max(UniqueCustomers) from (
select count(distinct userid) as UniqueCustomers from orders o where o.date between '2022-06-01' and '2022-07-01' group by rid 
) as temp
)
order by UniqueCustomers desc""",con=engine)

name  UniqueCustomers
0  dominos                2
1      kfc                2

#### -- 21. Identify users who have never placed an order with a specific partner (e.g., "Suresh")

In [14]:
pd.read_sql(""" select u.userid, u.name as UserName from users u 
where not exists (
select 1 from orders o join partner p on p.partnerid = o.partnerid where o.userid = u.userid and p.partnername = 'Suresh'
)
""",con=engine)

userid UserName
0       3  Vartika
1       6  Anupama
2       7  Rishabh

#### -- 22. Calculate the churn rate of users (percentage of users who stopped ordering after a specific month).

In [28]:
pd.read_sql(""" 
with ActiveBeforeMonth as (
select distinct o.userid  from orders o where o.date < '2022-07-01'
),
ActiveAfterMonth as (
select distinct o.userid from orders o where o.date > '2022-07-01'
),
ChurnCount as (
select b.userid from ActiveBeforeMonth b left join ActiveAfterMonth a on b.userid = a.userid where a.userid is NULL
)
select count(distinct c.userid) as ChurnedCount, count( distinct b.userid) as TotalActiveUsersBeforeMonth, 
count(distinct c.userid)*100/ count( distinct b.userid) as ChurnRate from ActiveBeforeMonth b left join ChurnCount c on b.userid = c.userid

""",con=engine)

ChurnedCount  TotalActiveUsersBeforeMonth  ChurnRate
0             1                            4       25.0

#### -- 23. Create a report showing the total revenue, number of orders, and average order value for each restaurant, grouped by month.

In [38]:
pd.read_sql(""" 
select r.name as RestoName, extract(month from o.date) as month, extract(year from o.date) as year, 
sum(o.amount) as Revenue, count(o.orderid) as TotalOrderPlaced, 
avg(o.amount) as AvgOrderValue from orders o join resto r on r.rid = o.rid group by 1,2,3 order by RestoName, month
""",con=engine)

RestoName  month  year  Revenue  TotalOrderPlaced  AvgOrderValue
0         box8      6  2022    480.0                 2          240.0
1         box8      7  2022    460.0                 2          230.0
2   China Town      6  2022    400.0                 1          400.0
3   China Town      7  2022   1050.0                 2          525.0
4      dominos      5  2022   1000.0                 2          500.0
5      dominos      6  2022    950.0                 1          950.0
6      dominos      7  2022   1100.0                 2          550.0
7   Dosa Plaza      5  2022    780.0                 3          260.0
8   Dosa Plaza      6  2022    400.0                 1          400.0
9   Dosa Plaza      7  2022    300.0                 1          300.0
10         kfc      5  2022    645.0                 2          322.5
11         kfc      6  2022    990.0                 3          330.0
12         kfc      7  2022   1935.0                 3          645.0

#### -- 24. Identify users who have placed consecutive orders on the same day.

In [43]:
pd.read_sql("""
 with consecutiveOrders as (
 select o.userid, o.orderid, o.date, lag(o.date) over(partition by userid order by date) as previousdate from orders o
 )
 select c.userid, u.name as UserName, c.date from users u join consecutiveOrders c on c.userid = u.userid where c.date = c.previousdate
""",con = engine)

Empty DataFrame
Columns: [userid, UserName, date]
Index: []

#### -- 25. Find the restaurant(s) with the highest number of orders containing only "Veg" items.

In [56]:
pd.read_sql(""" 
with OnlyVegOrders as (
select o.orderid, o.rid from orders o join orderdetails od on od.orderid = o.orderid join food f on f.fid = od.fid
group by 1,2 having sum(case when f.type <> 'Veg' then 1 else 0 end)=0
),
VegOrderCounts as (
select r.name as RestoName, count(v.orderid) as OrdersPlaced from OnlyVegOrders v join resto r on r.rid = v.rid group by 1
)
select RestoName , OrdersPlaced from VegOrderCounts where OrdersPlaced = (select max(OrdersPlaced) from VegORderCounts)
""",con=engine)

RestoName  OrdersPlaced
0  Dosa Plaza             5

## Final Conclusion

#### 1. Customer Insights
Users Who Have Never Ordered: The analysis identified 2 users (Anupama and Rishabh) who have never placed an order on Swiggy. 
This insight can be used to target these users with personalized offers or discounts to encourage their first order.

Favorite Foods: The most frequently ordered dishes include Choco Lava Cake and Chicken Wings, with Choco Lava Cake being the top favorite among users like Nitish, Khushboo, and Neha.

Ordering Patterns: Users like Nitish and Khushboo have placed multiple orders, indicating high engagement. 
Their favorite restaurants include kfc and dominos.

#### 2. Restaurant Performance
Top-Performing Restaurants: kfc emerged as the top-performing restaurant in June 2022, with the highest number of orders (3 orders) and significant revenue contributions.

Customer Satisfaction: While kfc had high sales, dominos had the lowest average customer satisfaction rating (1.0) despite generating high revenue. 
This indicates a need for improving service quality at dominos.

#### 3. Revenue Trends
Month-over-Month Growth: Swiggy experienced a 32.78% revenue growth from May to June 2022 and a 50.47% growth from June to July 2022.
This indicates a strong upward trend in business performance during this period.

Revenue Contribution: kfc contributed 34.03% of the total revenue, making it the highest revenue-generating restaurant, followed by dominos with 29.08%.

#### 4. Popular Cuisines and Dishes
Most Popular Cuisine: American cuisine (served by kfc) was the most popular, with 8 orders placed in the dataset.

Frequently Paired Dishes: The most frequently paired dishes were Choco Lava Cake and Chicken Wings, ordered together 5 times. 
This suggests that combo offers featuring these items could drive higher sales.

#### 5. Customer Retention
Churn Rate: The churn rate for users who stopped ordering after June 2022 was 25%, with 1 out of 4 active users not placing any orders in July. 
This highlights the need for retention strategies, such as loyalty programs or targeted discounts.

Loyal Customers: Users like Nitish, Khushboo, and Neha were identified as loyal customers, having placed more than 3 orders each. They were added to the loyalcustomers table and offered discounts based on their order value.

#### 6. Delivery Performance
Delivery Time Analysis: China Town had the highest average delivery time (70 minutes) in June 2022, indicating potential bottlenecks in their delivery process. This insight can help Swiggy optimize delivery logistics for better customer satisfaction.

#### 7. Veg vs. Non-Veg Preferences
Veg-Only Orders: Dosa Plaza had the highest number of orders containing only veg items (5 orders), making it a popular choice for vegetarian customers. This suggests that promoting veg-only restaurants could attract more vegetarian users.